File: 11/b.py


PRIME = 2 * 3 * 5 * 7 * 11 * 13 * 17 * 19

def MonkeyMove(monkey_id, object_worry):
	w = object_worry
	next_monkey = -1
	if (monkey_id == 0):
		w *= 17
		next_monkey = 2 if ((w % 2) == 0) else 6
	elif monkey_id == 1:
		w *= w
		next_monkey = 0 if ((w % 7) == 0) else 2

	elif monkey_id == 2:
		w += 7
		next_monkey = 7 if ((w % 13) == 0) else 6

	elif monkey_id == 3:
		w += 4
		next_monkey = 4 if ((w % 5) == 0) else 5

	elif monkey_id == 4:
		w += 5
		next_monkey = 1 if ((w % 3) == 0) else 5

	elif monkey_id == 5:
		w += 6
		next_monkey = 1 if ((w % 19) == 0) else 0

	elif monkey_id == 6:
		w *= 13
		next_monkey = 3 if ((w % 11) == 0) else 7

	elif monkey_id == 7:
		w += 2
		next_monkey = 4 if ((w % 17) == 0) else 3

	w = w % PRIME
	return {"next_owner" : next_monkey, "next_worry" : w}

ROUNDS = 10000
MONKE_COUNT = 8

monkeys = [
      [85, 79, 63, 72],
      [53, 94, 65, 81, 93, 73, 57, 92],
      [62, 63],
      [57, 92, 56],
      [67],
      [85, 56, 66, 72, 57, 99],
      [86, 65, 98, 97, 69],
      [87, 68, 92, 66, 91, 50, 68],
]

inspects = [0 for i in range(MONKE_COUNT)]
ct = 0 
while (ct < ROUNDS):
	ct += 1
	print(ct)

	for i in range(MONKE_COUNT):
		for item in monkeys[i]:
			inspects[i]+=1
			res = MonkeyMove(i, item)
			monkeys[res["next_owner"]].append(res["next_worry"])
			monkeys[i] = []

inspects.sort()
print(inspects)
print (inspects[-1] * inspects[-2])

File: 11/b.cpp
#include <bits/stdc++.h>
#define TRACE(x) cout << "[" << #x << ": " << x << "]" << endl;
#define TRACE2(x, y)                                                           \
  cout << "[" << #x << ": " << x << "]"                                        \
       << " || "                                                               \
       << "[" << #y << ": " << y << "]" << endl;
#define int long long
using namespace std;

const string filename = "input.txt";
const int PRIME = 2 * 3 * 5 * 7 * 11 * 13 * 17 * 19;

struct TurnResult {
  int next_owner;
  int next_worry;
};

struct Monke {
  vector<int> items;
  void add_item(int item) { items.push_back(item); }
};

TurnResult MonkeyMove(int monkey_id, int object_worry) {
  int w = object_worry;
  int next_monkey = -1;

  switch (monkey_id) {
  case 0:
    w *= 17;

    next_monkey = ((w % 2) == 0) ? 2 : 6;
    break;

  case 1:
    w *= w;

    next_monkey = ((w % 7) == 0) ? 0 : 2;
    break;

  case 2:
    w += 7;

    next_monkey = ((w % 13) == 0) ? 7 : 6;
    break;

  case 3:
    w += 4;

    next_monkey = ((w % 5) == 0) ? 4 : 5;
    break;

  case 4:
    w = (w + 5);

    next_monkey = ((w % 3) == 0) ? 1 : 5;
    break;

  case 5:
    w = (w + 6);

    next_monkey = ((w % 19) == 0) ? 1 : 0;
    break;

  case 6:
    w = (w * 13);

    next_monkey = ((w % 11) == 0) ? 3 : 7;
    break;

  case 7:
    w = (w + 2);

    next_monkey = ((w % 17) == 0) ? 4 : 3;
    break;

  default:
    assert(0);
    break;
  }

  w %= PRIME;
  return (TurnResult){.next_owner = next_monkey, .next_worry = w};
}

vector<int> parse(string &line, char delim) {
  vector<int> res;
  string token;
  int it = 0;
  while (it < line.size()) {
    while (it < line.size() && line[it] != delim) {
      token += line[it];
      it++;
    }

    if (line[it] == ',')
      it++;
    res.push_back(stoi(token));
    token.clear();
  }
  return res;
}

signed main() {
  ifstream input(filename);
  assert(input.is_open());
  int ROUNDS = 10000;
  const int MONKE_COUNT = 8;

  Monke monkeys[MONKE_COUNT] = {
      (Monke){.items = vector<int>{85, 79, 63, 72}},
      (Monke){.items = vector<int>{53, 94, 65, 81, 93, 73, 57, 92}},
      (Monke){.items = vector<int>{62, 63}},
      (Monke){.items = vector<int>{57, 92, 56}},
      (Monke){.items = vector<int>{67}},
      (Monke){.items = vector<int>{85, 56, 66, 72, 57, 99}},
      (Monke){.items = vector<int>{86, 65, 98, 97, 69}},
      (Monke){.items = vector<int>{87, 68, 92, 66, 91, 50, 68}},
  };

  int inspects[MONKE_COUNT] = {0};

  while (ROUNDS--) {
    for (int i = 0; i < MONKE_COUNT; i++) {
      for (auto item : monkeys[i].items) {
        inspects[i]++;
        auto res = MonkeyMove(i, item);
        // TRACE2(res.next_owner, res.next_worry);
        monkeys[res.next_owner].add_item(res.next_worry);
      }
      monkeys[i].items.clear();
    }
  }

  sort(inspects, inspects + MONKE_COUNT);
  for (int i = 0; i < MONKE_COUNT; i++) {
    TRACE2(i, inspects[i]);
  }

  int kek = inspects[MONKE_COUNT - 1] * 1LL * inspects[MONKE_COUNT - 2];
  TRACE(kek);

  return 0;
}
